src/libxively/gateway/xi_gw_glue_layer.c

```diff
@@ -0,0 +1,123 @@
+/* Copyright (c) 2003-2018, LogMeIn, Inc. All rights reserved.
+ *
+ * This is part of the Xively C Client library,
+ * it is licensed under the BSD 3-Clause license.
+ */
+
+#include <xi_gw_glue_layer.h>
+#include <xi_layer_api.h>
+#include <xively_mqtt.h>
+#include <xively.h>
+#include <xi_types.h>
+
+static void
+_xi_message_arrived_on_tunnel_callback( xi_context_handle_t in_context_handle,
+                                        xi_sub_call_type_t call_type,
+                                        const xi_sub_call_params_t* const params,
+                                        xi_state_t state,
+                                        void* user_data )
+{
+    printf( "--- %s ---, state: %d, user_data: %p\n", __FUNCTION__, state, user_data );
+
+    XI_UNUSED( in_context_handle );
+    XI_UNUSED( call_type );
+
+    if ( XI_MQTT_SUBSCRIPTION_SUCCESSFULL == state )
+    {
+        /* suback */
+    }
+    else if ( XI_STATE_OK == state )
+    {
+        /* PUBLISH arrived on tunnel */
+        xi_data_desc_t* data = xi_make_desc_from_buffer_copy(
+            params->message.temporary_payload_data,
+            params->message.temporary_payload_data_length );
+
+        XI_PROCESS_PULL_ON_THIS_LAYER( user_data, data, state );
+    }
+}
+
+void _xi_publish_result_callback( xi_context_handle_t in_context_handle,
+                                  void* user_data,
+                                  xi_state_t state )
+{
+    printf( "--- --- --- %s ---, state: %d, user_data: %p\n", __FUNCTION__, state,
+            user_data );
+    XI_UNUSED( in_context_handle );
+
+    XI_PROCESS_PUSH_ON_THIS_LAYER( user_data, NULL,
+                                   ( XI_STATE_OK == state ) ? XI_STATE_WRITTEN : state );
+}
+
+xi_state_t xi_gw_glue_layer_init( void* context, void* data, xi_state_t in_out_state )
+{
+    XI_LAYER_FUNCTION_PRINT_FUNCTION_DIGEST();
+
+    /* subscribing main Xively Client to edge device specific tunnel topic */
```

Comment:
this comment is a bit confusing.  Can you reword it? :'/

Reply:
simplified a bit: `/* subscribing to edge device's tunnel topic */`